aten/src/ATen/cuda/CUDAFuture.h

```diff
@@ -132,25 +119,21 @@ struct TORCH_CUDA_API CUDAFuture : at::ivalue::Future {
   // is first marked completed.
   std::vector<std::reference_wrapper<const at::DataPtr>> dataPtrs_;
 
- private:
-  DataPtrExtractor dataPtrExtractor_;
-  std::mutex dataPtrExtractorMutex_;
-
   // FIXME This too is protected so that it can be used by FutureNCCL. Please
   // undo that once FutureNCCL is dropped in favor of a "vanilla" CUDAFuture.
  protected:
   std::vector<std::reference_wrapper<const at::DataPtr>> extractDataPtrs(
       const at::IValue& value) {
-    std::unique_lock<std::mutex> lock(dataPtrExtractorMutex_);
+    at::IValue::HashAliasedIValues sub_values;
+    // Prefer getSubValues() over visit() as the latter is a silent no-op for
+    // some unsupported types, whereas the former at least fails loudly.
+    value.getSubValues(sub_values);
```

Comment:
Would I be correct if I assume this gonna extract all nested ivalues in the given ivalue object? Curious, does it make sense to let `CUDAFuture::markCompleted` (or other APIs) also take a vector of tensors? Asking because RPC already have access to the tensor vector and hence does not need to recursively visit all nested values again.

Reply:
> Would I be correct if I assume this gonna extract all nested ivalues in the given ivalue object?

Yes, indeed.

Note that `markCompleted` is also called inside `then()` with the result of the *user-provided* callback, which can return an arbitrary value. (Unless we explicitly lock that down and force users to also always return lists of tensors, is this what you're suggesting?)

Reply:
> Unless we explicitly lock that down and force users to also always return lists of tensors, is this what you're suggesting?

No, I am thinking about an additional fast path that can take a list of tensors. So users can either pass in an ivalue or ivalue + list of tensors to avoid traverse the ivalue again. Not sure if this makes sense.

Reply:
I see, yes, that could be an option. We could either add such a method to ivalue::Future, or we could add it just to CUDAFuture: this second option has the problem that the new method wouldn't be "accessible" anymore once we downcast the CUDAFuture to an ivalue::Future, but that might not matter, because markCompleted is only called by whoever _creates_ the object, and they have access to the specialized subclass pointer.